aten/src/ATen/native/cudnn/BatchNorm.cpp

```diff
@@ -195,41 +201,103 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
-      handle, mode, &one, &zero,
-      idesc.desc(), input->data_ptr(),
-      idesc.desc(), output->data_ptr(),
-      wdesc.desc(),
-      weight->data_ptr(),
-      bias->data_ptr(),
-      exponential_average_factor,
-      at::maybe_data_ptr(running_mean),
-      at::maybe_data_ptr(running_var),
-      epsilon,
-      save_mean.data_ptr(),
-      save_var.data_ptr()));
+        handle,
+        mode,
+        &one,
+        &zero,
+        idesc.desc(),
+        input->data_ptr(),
+        idesc.desc(),
+        output->data_ptr(),
+        wdesc.desc(),
+        weight->data_ptr(),
+        bias->data_ptr(),
+        exponential_average_factor,
+        at::maybe_data_ptr(running_mean),
+        at::maybe_data_ptr(running_var),
+        epsilon,
+        save_mean.data_ptr(),
+        save_var.data_ptr()));
 #endif // CUDNN_VERSION >= 7400
   } else {
-    reserve = at::empty({0}, input->options().dtype(kByte));
     AT_CUDNN_CHECK(cudnnBatchNormalizationForwardInference(
-      handle, mode, &one, &zero,
-      idesc.desc(), input->data_ptr(),
-      idesc.desc(), output->data_ptr(),
-      wdesc.desc(),
-      weight->data_ptr(),
-      bias->data_ptr(),
-      running_mean->data_ptr(),
-      running_var->data_ptr(),
-      epsilon));
+        handle,
+        mode,
+        &one,
+        &zero,
+        idesc.desc(),
+        input->data_ptr(),
+        idesc.desc(),
+        output->data_ptr(),
+        wdesc.desc(),
+        weight->data_ptr(),
+        bias->data_ptr(),
+        running_mean->data_ptr(),
+        running_var->data_ptr(),
+        epsilon));
   }
 
   // save_mean and save_var can be undefined
   // If this causes problems, we can initialize them to empty tensors
   // of the correct type
-  return std::tuple<Tensor, Tensor, Tensor, Tensor>{output_t, save_mean, save_var, reserve};
+  return std::tuple<Tensor&, Tensor&, Tensor&, Tensor&>{
+      output_t, save_mean, save_var, reserve};
 }
 
+std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
+    const Tensor& input_t,
+    const Tensor& weight_t,
+    const c10::optional<Tensor>& bias_t_opt,
+    const c10::optional<Tensor>& running_mean_t_opt,
+    const c10::optional<Tensor>& running_var_t_opt,
+    bool training,
+    double exponential_average_factor,
+    double epsilon) {
+  TensorArg input{input_t, "input", 1};
+  auto output_t =
+      at::empty_like(*input, input->options(), input->suggest_memory_format());
+
+  auto handle = getCudnnHandle();
+  auto dataType = getCudnnDataType(*input);
+  TensorDescriptor idesc{*input, 4}; // input descriptor
+  Tensor save_mean, save_var, reserve;
+  cudnnBatchNormMode_t mode = getCudnnBatchNormMode(
+      training, input->suggest_memory_format(), input->dim());
+
+  if (training) {
+    int64_t num_features = input_t.size(1);
+    save_mean = at::empty({num_features}, weight_t.options());
+    save_var = at::empty({num_features}, weight_t.options());
+
+#if CUDNN_VERSION >= 7400
```

Comment:
`reserve` is not allocated for `#else` or `!training`?

Reply:
my reasoning here was three fold:

1) a reserve buffer is indeed not used, neither for cudnn < 7.4 nor in forward pass  
2) up above we have this

>  // save_mean and save_var can be undefined
  // If this causes problems, we can initialize them to empty tensors
  // of the correct type
  
 thus reasoning by analogy...

3) return is ultimately consumed in [Tensor batch_norm](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L538)
```
return std::get<0>(at::_batch_norm_impl_index(input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled));
```
which discards. 

but it was just my intuition 🤷. if you think this would break something i will certainly put it back.